src/ctrans.c

```diff
@@ -320,142 +354,282 @@ int matmulti(_float *val, int n, _float mat[][3], int skip){
-		       grid_start, grid_stop, data_size, grid_nsteps, norm_data);
-  
+
+  meanout = (PyArrayObject*)PyArray_SimpleNew(3, dims, NPY_DOUBLE);
+  if(!meanout){
+    PyErr_SetString(PyExc_MemoryError, "Could not allocate memory (meanout)");
+    goto error;
+  }
+ 
+  int retval;
+  retval = c_grid3d((double*)PyArray_DATA(gridout), (unsigned long*)PyArray_DATA(Nout),
+                    (double*)PyArray_DATA(meanout), (double*)PyArray_DATA(stderror), 
+                    (double*)PyArray_DATA(gridI), &n_outside,
+		                grid_start, grid_stop, (unsigned long)data_size, grid_nsteps, 1);
+
+  if(retval){
+    // We had a runtime error
+    PyErr_SetString(PyExc_RuntimeError, "Gridding process failed to run");
+    goto error;
+  }
+
   Py_XDECREF(gridI);
-  return Py_BuildValue("NNNl", gridout, Nout, standarderror, n_outside);
-  
- cleanup:
+  return Py_BuildValue("NNNNl", gridout, meanout, Nout, stderror, n_outside);
+
+error:
   Py_XDECREF(gridI);
   Py_XDECREF(gridout);
+  Py_XDECREF(meanout);
   Py_XDECREF(Nout);
-  Py_XDECREF(standarderror);
+  Py_XDECREF(stderror);
   return NULL;
 }
 
-unsigned long c_grid3d(double *dout, unsigned long *nout, double *standarderror, double *data,
-		       double *grid_start, double *grid_stop, int max_data,
-		       int *n_grid, int norm_data){
-  int i;
-  unsigned long *nout_ptr;
-  double *dout_ptr;
-  double *data_ptr;
-  
-  double *Mk = NULL;
-  double *Qk = NULL;
-  int grid_size = 0;
-
-  double pos_double[3];
-  double grid_len[3], grid_step[3];
-  int grid_pos[3];
-  int pos = 0;
-  unsigned long n_outside = 0;
+int c_grid3d(double *dout, unsigned long *nout, double *mout,
+             double *stderror, double *data, unsigned long *n_outside,
+             double *grid_start, double *grid_stop, unsigned long max_data,
+             unsigned long *n_grid, int norm){
+  unsigned long i, j;
+  unsigned long grid_size = 0;
+  double grid_len[3];
+  unsigned long stride;
+  unsigned long *nthread = NULL;
 	
   // Some useful quantities
 
   grid_size = n_grid[0] * n_grid[1] * n_grid[2];
+  for(i = 0;i < 3; i++){
+    grid_len[i] = grid_stop[i] - grid_start[i];
+  }
+
+  // If we do this with threads .. we can do map reduce
+
+#ifdef USE_THREADS
+  pthread_t thread[MAX_THREADS];
+#endif
+
+  gridderThreadData threadData[MAX_THREADS];
 
   // Allocate arrays for standard error calculation
-  
-  if(standarderror){
-    Mk = (double*)malloc(sizeof(double) * grid_size);
-    if(!Mk){
-      return n_outside;
+ 
+  for(i=0;i<_n_threads;i++){
+    threadData[i].Mk = NULL;
+    threadData[i].Qk = NULL;
+    threadData[i].dout = NULL;
+    threadData[i].nout = NULL;
+  }
+
+  nthread = (unsigned long *)malloc(sizeof(unsigned long) * grid_size);
+  if(!nthread){
+    goto error;
+  }
+  for(i=0;i<grid_size;i++){
+    nthread[i] = 0;
+  }
+
+  stride = max_data / _n_threads;
+  for(i=0;i<_n_threads;i++){
+    threadData[i].Qk = (double*)malloc(sizeof(double) * grid_size);
```

Comment:
There may be some memory leaking of those arrays unless they are freed at a top level.